File: apps/android/core/src/main/kotlin/app/trackevolution/core/model/Entitlement.kt
package app.trackevolution.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * `entitlement` on `GET /api/me` — the server's answer to "is this account Pro"
 * (NS-32). The server owns it: the app never decides tier from a local
 * purchase, and offline the last fetched value stands.
 */
@Serializable
public data class Entitlement(
    val tier: Tier,
    /**
     * Where the entitlement came from: the store that sold it, `legacy` for the
     * paid-app grant, or null for an account that never had one. A lapsed
     * subscriber keeps their source so "Manage" can still target the right store.
     */
    val source: Source? = null,
    /** Epoch ms. Null for free and for legacy (which has no expiry). */
    @SerialName("expires_at") val expiresAt: Long? = null,
    /** Null when unknown (free, legacy); false when the store says it won't renew. */
    @SerialName("auto_renew") val autoRenew: Boolean? = null,
) {
    @Serializable
    public enum class Tier {
        @SerialName("free") FREE,
        @SerialName("pro") PRO,
    }

    @Serializable
    public enum class Source {
        @SerialName("legacy") LEGACY,
        @SerialName("apple") APPLE,
        @SerialName("google") GOOGLE,
    }

    /** What a client-side gate decides: carry on, or show the paywall instead. */
    public enum class Gate { PROCEED, PAYWALL }

    /**
     * The tier predicates — the port of `public/js/entitlement.js`, under the
     * same names, pinned case for case by `contracts/logic/entitlement.json`
     * (the fixture the iOS Kit asserts against too, so the two ports are checked
     * against the web implementation rather than against each other).
     *
     * Every one of these takes the entitlement object and **never consults the
     * clock**: expiry is the server's call, made when it answered `/api/me`.
     * Offline, the cached answer stands — a driver who was Pro at the last sync
     * records (NS-32 rule 5) — which is why a cached `expires_at` already in the
     * past on the phone's clock is still Pro here.
     */
    public companion object {
        /** What a signed-out or never-fetched client has: free. */
        public val FREE: Entitlement = Entitlement(tier = Tier.FREE)

        /**
         * On since phase D. Phase C shipped the purchase flow, the paywall and
         * the Settings row dark behind this one constant; flipping it is what
         * turned the recorder and import gates on. The gates take the flag as a
         * parameter defaulting to this, so the tests exercise both values.
         */
        public const val GATES_ENABLED: Boolean = true

        public const val APPLE_MANAGE_URL: String = "https://apps.apple.com/account/subscriptions"
        public const val GOOGLE_MANAGE_URL: String =
            "https://play.google.com/store/account/subscriptions?package=app.trackevolution"

        /** An entitlement, or nothing at all (signed out, never fetched) — the latter is free, never Pro. */
        public fun isPro(entitlement: Entitlement?): Boolean = entitlement?.tier == Tier.PRO

        /** The GPS lap recorder, live timing and predictive delta. */
        public fun canRecord(entitlement: Entitlement?): Boolean = isPro(entitlement)

        // Telemetry import is **free** and has no predicate, on purpose: there
        // is no decision to make, and a `canImport` that always answered true
        // would read as a gate someone had forgotten to wire. An import yields
        // lap times, the racing line and the car metrics for free; the per-lap
        // channel arrays it also writes are the Pro half, withheld by the
        // server — see [canViewChannels].

        /**
         * Channel graphs, the lap delta chart, the two-lap compare and sector
         * splits all read `channels`, which the server strips for a free account
         * (rule 4); this only decides whether the resulting empty state carries
         * the paywall copy.
         */
        public fun canViewChannels(entitlement: Entitlement?): Boolean = isPro(entitlement)

        /** Garage consumables, the setup notebook and year in review. */
        public fun canUseGarage(entitlement: Entitlement?): Boolean = isPro(entitlement)
        public fun canUseSetups(entitlement: Entitlement?): Boolean = isPro(entitlement)
        public fun canViewYearInReview(entitlement: Entitlement?): Boolean = isPro(entitlement)

        /**
         * The web's two-event lap overlay. Ported for name parity with
         * `public/js/entitlement.js` and the shared fixture, though no Android
         * screen reads it — the overlay is web-only (docs/specs/native/README.md).
         */
        public fun canCompareEvents(entitlement: Entitlement?): Boolean = isPro(entitlement)

        /**
         * Where "Manage subscription" goes, by the store that sold it. Legacy has
         * no subscription to manage and free has nothing yet — both null. A lapsed
         * subscriber keeps their source, so Manage still targets the right store.
         */
        public fun manageUrl(entitlement: Entitlement?): String? = when (entitlement?.source) {
            Source.APPLE -> APPLE_MANAGE_URL
            Source.GOOGLE -> GOOGLE_MANAGE_URL
            Source.LEGACY, null -> null
        }

        /**
         * One line for Settings: "Pro · renews Mar 4, 2027", "Pro · ends …",
         * "Pro · lifetime", "Free". [fmtDate] is injected, as in the JS, so the
         * rule stays locale-free and the fixture can pin the wording without the
         * date text; the default matches the JS default (an ISO date, UTC).
         */
        public fun entitlementSummary(
            entitlement: Entitlement?,
            fmtDate: (Long) -> String = ::isoDate,
        ): String {
            if (!isPro(entitlement)) return "Free"
            val expiresAt = entitlement!!.expiresAt
            if (entitlement.source == Source.LEGACY || expiresAt == null) return "Pro · lifetime"
            val when_ = fmtDate(expiresAt)
            return if (entitlement.autoRenew == false) "Pro · ends $when_" else "Pro · renews $when_"
        }

        /**
         * The recorder's start gate (rule 5): a paywall sheet, never a disabled
         * button, and only when the gates are on. The only client-side gate
         * there is — importing is free. Pure so the decision stays tested with
         * the flag injected, whatever [GATES_ENABLED] currently is.
         */
        public fun recordGate(entitlement: Entitlement?, gatesEnabled: Boolean = GATES_ENABLED): Gate =
            if (!gatesEnabled || canRecord(entitlement)) Gate.PROCEED else Gate.PAYWALL

        /** `new Date(ms).toISOString().slice(0, 10)` — the JS module's default. */
        private fun isoDate(ms: Long): String =
            java.time.Instant.ofEpochMilli(ms).atOffset(java.time.ZoneOffset.UTC).toLocalDate().toString()
    }
}

/**
 * Every billing write (`POST /api/billing/apple`, `/google`, and the two legacy
 * claims) answers with the fresh entitlement so the client can update without a
 * second round trip.
 */
@Serializable
public data class BillingResponse(
    val ok: Boolean,
    val entitlement: Entitlement,
)

File: public/js/entitlement.js
// The tier predicates (NS-32 requirement 8), the one place the client-side
// half of the tier table lives. Every decision a client makes from
// `entitlement` on GET /api/me goes through these, and the iOS Kit and Android
// :core carry ports under the same names, pinned to
// contracts/logic/entitlement.json — so add a predicate here first.
//
// Offline, the cached /api/me answer stands: a driver who was Pro at the last
// sync records (rule 5). That is why these take the entitlement object and
// never consult the clock — expiry is the server's call, made when it answered.

// Free: what the logbook is today. Pro: the analysis. The web and both apps
// gate at the point of use, never on a write.
export const FREE_ENTITLEMENT = Object.freeze({
  tier: "free",
  source: null,
  expires_at: null,
  auto_renew: null,
});

// An entitlement object, or nothing at all (not signed in, never fetched) —
// the latter is free, never Pro.
export function isPro(entitlement) {
  return !!entitlement && entitlement.tier === "pro";
}

// The GPS lap recorder, live timing and predictive delta (native only).
export const canRecord = (entitlement) => isPro(entitlement);

// Telemetry import is **free** on every client and has no predicate, on
// purpose: there is no decision to make, and a `canImport` that always answered
// true would read as a gate that had been forgotten. What an import yields for
// free is lap times, the racing line and the car metrics; the per-lap channel
// arrays it also writes are the Pro half, and the server is what withholds them
// (`stripProFields`) — see canViewChannels.

// The three traces a free account keeps (#264): speed, throttle and brake —
// the ones any driver can read at a glance. The server enforces this
// (`FREE_CHANNELS` in src/lib/entitlement.ts strips every other gridded
// channel and every per-lap scalar for a free account); the client-side
// predicate only decides what to *say* about a channel that isn't there.
export const FREE_CHANNELS = Object.freeze(["speed", "throttle", "brake"]);
export const canViewChannel = (entitlement, key) => FREE_CHANNELS.includes(key) || isPro(entitlement);

// The Pro half of the channel panel: every channel beyond FREE_CHANNELS
// (steering, RPM, lateral G, yaw, gear, the limit marks, the Car tab), the
// lap delta chart, sector splits and the two-lap compare. The server strips
// the channels themselves (rule 4); the delta and the sectors derive from the
// free speed trace, so for those this predicate is the gate, and it is also
// what decides whether the panel carries the locked note.
export const canViewChannels = (entitlement) => isPro(entitlement);

// Garage consumables, the setup notebook and year in review.
export const canUseGarage = (entitlement) => isPro(entitlement);
export const canUseSetups = (entitlement) => isPro(entitlement);
export const canViewYearInReview = (entitlement) => isPro(entitlement);

// The two-event lap overlay — one track, two events, lap by lap. Separate from
// canViewChannels because it reads no channel data at all, only lap times, so
// the two would move independently if the tier boundary ever did.
export const canCompareEvents = (entitlement) => isPro(entitlement);

// Where "Manage subscription" goes, by the store that sold it. Legacy has no
// subscription to manage; free has nothing yet — both return null.
export const APPLE_MANAGE_URL = "https://apps.apple.com/account/subscriptions";
export const GOOGLE_MANAGE_URL =
  "https://play.google.com/store/account/subscriptions?package=app.trackevolution";

export function manageUrl(entitlement) {
  if (!entitlement) return null;
  if (entitlement.source === "apple") return APPLE_MANAGE_URL;
  if (entitlement.source === "google") return GOOGLE_MANAGE_URL;
  return null;
}

// One line for Settings: "Pro · renews Mar 4, 2027", "Pro · ends Mar 4, 2027",
// "Pro · lifetime", "Free". `fmtDate` is injected so the module stays free of
// locale/Intl and the ports can pin the shape without the date text.
export function entitlementSummary(entitlement, fmtDate = (ms) => new Date(ms).toISOString().slice(0, 10)) {
  if (!isPro(entitlement)) return "Free";
  if (entitlement.source === "legacy" || entitlement.expires_at == null) return "Pro · lifetime";
  const when = fmtDate(entitlement.expires_at);
  return entitlement.auto_renew === false ? `Pro · ends ${when}` : `Pro · renews ${when}`;
}
